src/canister_tests/src/certificate_validation.rs

```diff
@@ -0,0 +1,147 @@
+// TODO: certificate validation should be its own library
+
+use crate::certificate_validation::ValidationError::{
+    AssetHashMismatch, AssetPathLookupFailed, MalformedCertificate, TreeHashCertifiedDataMismatch,
+    WitnessLookupFailed,
+};
+use candid::types::ic_types::hash_tree::LookupResult;
+use flate2::read::GzDecoder;
+use ic_agent::agent::http_transport::ReqwestHttpReplicaV2Transport;
+use ic_agent::hash_tree::HashTree;
+use ic_agent::{lookup_value, Agent, AgentError};
+use ic_crypto_utils_threshold_sig_der::public_key_to_der;
+use ic_state_machine_tests::{CanisterId, ThresholdSigPublicKey};
+use regex::Regex;
+use serde::Deserialize;
+use sha2::{Digest, Sha256};
+use std::io::Read;
+
+// The limit of a buffer we should decompress ~10mb.
+const MAX_CHUNK_SIZE_TO_DECOMPRESS: usize = 1024;
+const MAX_CHUNKS_TO_DECOMPRESS: u64 = 10_240;
+
+const CERTIFICATE_REGEX: &'static str = "^certificate=:([^:]*):, tree=:([^:]*):$";
+
+#[derive(Debug)]
+pub enum ValidationError {
+    MalformedCertificate { message: String },
+    CertificateValidationFailed { inner: AgentError },
+    WitnessLookupFailed { inner: AgentError },
+    TreeHashCertifiedDataMismatch,
+    AssetPathLookupFailed,
+    AssetHashMismatch,
+}
+
+/// Validates asset certification according to the HTTP gateway specification:
+/// https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-gateway-certification
+pub fn validate_certification(
+    ic_certificate: &str,
+    canister_id: CanisterId,
+    uri_path: &str,
+    body: &[u8],
+    encoding: Option<&str>,
+    root_key: ThresholdSigPublicKey,
+) -> Result<(), ValidationError> {
+    let cert = parse_header_value(ic_certificate, 1)?;
+    let tree: HashTree = parse_header_value(ic_certificate, 2)?;
+    let agent = initialize_agent(&root_key);
+
+    if let Err(err) = agent.verify(&cert, canister_id.get().0, false) {
+        return Err(ValidationError::CertificateValidationFailed { inner: err });
+    }
+
+    let certified_data_path = vec![
+        "canister".into(),
+        canister_id.into(),
+        "certified_data".into(),
+    ];
+    let witness = match lookup_value(&cert, certified_data_path) {
+        Ok(witness) => witness,
+        Err(err) => {
+            return Err(WitnessLookupFailed { inner: err });
+        }
+    };
+
+    if *witness != tree.digest() {
+        return Err(TreeHashCertifiedDataMismatch);
+    }
+
+    let asset_path = ["http_assets".into(), uri_path.into()];
+    let tree_sha = match tree.lookup_path(&asset_path) {
+        LookupResult::Found(v) => v,
+        _ => match tree.lookup_path(&["http_assets".into(), "/index.html".into()]) {
+            LookupResult::Found(v) => v,
+            _ => {
+                return Err(AssetPathLookupFailed);
+            }
+        },
+    };
+
+    let body_sha = decode_body_to_sha256(body, encoding).unwrap();
```

Comment:
maybe add a comment about how II doesn't follow the spec since gzipped content has the hash of the gzipped content?